clang/include/clang/CIR/Dialect/IR/CIROps.td

```diff
@@ -1847,45 +1863,111 @@ def SwitchOp : CIR_Op<"switch",
 
-    Each case region must be explicitly terminated.
+    The `cir.case` operations doesn't have to be in the region of `cir.switch` directly. However,
+    when all the `cir.case` operations lives in the region of `cir.switch` directly and there is
+    no other operations except the ending `cir.yield` operation in the region of `cir.switch` directly,
+    we call the `cir.switch` operation is in a simple form. Users can use
+    `bool isSimpleForm(llvm::SmallVector<CaseOp> &cases)` member function to detect if the `cir.switch` operation
+    is in a simple form. The simple form makes analysis easier to handle the `cir.switch` operation
+    and makes the boundary to give up pretty clear.
 
-    Examples:
+    To make the simple form as common as possible, we will attach the operations corresponding to the statements
+    that lives between top level cases into the closest `cir.case` operation.
 
-    ```mlir
-    cir.switch (%b : i32) [
-      case (equal, 20) {
-        ...
-        cir.yield break
-      },
-      case (anyof, [1, 2, 3] : i32) {
-        ...
-        cir.return ...
+    For example,
+
+    ```
+    switch(int cond) {
+      case 4:
+        a++;
+
+      b++;
+      case 5;
+        c++;
+
+      ...
+    }
+    ```
+
+    The statement `b++` is not a sub-statement of the case statement `case 4`. But to make the generated `cir.switch`
+    a simple form, we will attach the statement `b++` into the closest `cir.case` operation. So that the generated code
+    will be like:
+
+    ```
+    cir.switch(int cond) {
+      cir.case(equal, 4) {
+        a++;
+        b++;
+        cir.yield
       }
-      case (range, [10, 15]) {
-        ...
-        cir.yield break
-      },
-      case (default) {
-        ...
-        cir.yield fallthrough
+      cir.case(equal, 5) {
+        c++;
+        cir.yield
       }
-    ]
+      ...
+    }
+    ```
+
+    For the same reason, we will hoist the case statement as the substatement of another case statement so that they will
+    be in the same level. For example,
+
+    ```
+    switch(int cond) {
+      case 4:
+      default;
+      case 5;
+        a++;
+      ...
+    }
+    ```
+
+    will be generated as
+
+    ```
+    cir.switch(int cond) {
+      cir.case(equal, 4) {
+        cir.yield
+      }
+      cir.case(default) {
+        cir.yield
+      }
+      cir.case(equal, 5) {
+        a++;
+        cir.yield
+      }
+      ...
+    }
+    ```
+
+    So that, the `cir.switch` may not be in a simple form if only any of the following cases are true:
+    - There are case statements of the switch statement lives in other scopes other than the top level compound statement scope.
+      Note that a case statement itself doesn't form a scope.
+    - The sub-statement of the switch statement is not a compound statement.
+    - There are codes before the first case statement. For example,
+
+    ```
+    switch(int cond) {
```

Comment:
Since you already went all the way to provide cool examples, how about you add some CIR for this C++?